design/tls-client-verification.md

```diff
@@ -0,0 +1,231 @@
+# Client certificate validation (mTLS)
+
+Status: Draft
+
+## Goals
+
+- Allow client certificate validation (mTLS) on contour /performed by Envoy)
+- Allow various ways for client certificate validation; Spki, Hash and CA
+- Forward client certificate details to the backend (in the XFCC header)
+- Document mTLS configuration and client certificate detail forwarding.
+
+## Non Goals
+
+- Support for k8s ingress documents.
+- Repeat configuration details described in Envoy documentation (make references)
+
+## Background
+
+Client certificate validation (mTLS) is supported by Envoy. It should
+be possible for `contour` users to utilize this feature.
+
+
+## Implementation phases
+
+The most urgent and also the most straight-forward feature is to
+provide client certificate validation. For some backend applications
+(but likely not for all) the client cerificate details are needed. So
+the proposal is to divide the mTLS implementation in two phases;
+
+1. Add client certificate validation but with hard-coded client
+   certification detail forwarding (XFCC header)
+
+2. Add configuration of what is to be forwarded in the XFCC header
+
+Configuration of the XFCC headed can be added later without make any
+backward-incompatible changes.
+
+The main problem with XFCC is that it is not really part of the tls
+configuration, but rather a backend http connection setting. This
+makes me believe that some discussions must be conducted before the
+configuration is settled. But in the meanwhile implementation the more
+urgent mTLS validation must not be stalled.
+
+The only one of the "other" ingress controllers that mention client
+certificate details forwarding I have found is `traefik`;
+
+https://docs.traefik.io/v2.0/middlewares/passtlsclientcert/
+
+For some reason they do not use the "standard" XFCC header. Never the
+less is may be wise to mimic this approach in `contour`, for instance
+herd-code forwaring of the entire certificate rather than the hash in
+phase 1.
+
+
+## High-Level Design
+
+At a high level I propose the following:
+
+1. A new record "clientValidation" is added in spec.virtualhost.tls
```

Comment:
You lost me here. The previous section talked about XFCC being between the end user and and backend application server. Now this is talking about a change to the tls settings for the whole virtual host. Is this a problem with the problem statement? Is this a security issue? If so what is the threat model? Is this an authentication issue? It's not clear to me at the moment.

Reply:
I think you may be right. I will remove the XFCC header from the proposal since it may be confusing. It can be handled separately later on in another issue.

Reply:
Hey @uablrek, just taking a look here as well. 

I'm curious about @davecheney's comment earlier about if this is a vhost setting or a per route setting. Could I have each route in my cluster use a different cert based upon the upstream it's calling?

Reply:
@stevesloka I asked the very same question in the Envoy slack :smile: The ansver is; No client certificate validation is per vhost, not for routes. The reason is that the route is in the data, but the client validation is performed before data is decrypted.

Reply:
For the XFCC header on the other hand it seems possible to forward different certificate details to dirrefent routes. But as @davecheney it is much better to add XFCC support as a different item, so I do not mention it in the design document any more.

Reply:
Ok so to clarify, is this looking to mTLS a public client calling into Envoy with a backend service? Or to secure communication from Envoy to a backend service inside the cluster?

Reply:
This is *not* about cluster internal Envoy-backend communication. It is mTLS used to validate an external client accessing the service through contour/envoy from outside the cluster.